In [1]:
!pip install -q pyomo

In [2]:
from pyomo.environ import * 

In [3]:
import numpy as np

In [4]:
import pandas as pd

In [5]:
data_txtfile = np.loadtxt('lab6_ex2.txt')
print(data_txtfile)

[[21. 25. 18. 19. 23. 24. 16. 17. 21. 23. 19. 21.]
 [19. 22. 19. 19. 23. 21. 15. 20. 16. 17. 17. 24.]
 [21. 19. 20. 18. 22. 24. 14. 16. 18. 16. 20. 24.]
 [19. 22. 17. 17. 22. 22. 18. 18. 20. 21. 23. 23.]
 [18. 22. 15. 15. 23. 23. 21. 25. 24. 19. 21. 23.]
 [23. 19. 19. 14. 24. 21. 23. 21. 20. 20. 22. 19.]
 [21. 16. 27. 17. 26. 27. 16. 19. 24. 20. 20. 23.]
 [22. 17. 21. 15. 23. 24. 24. 20. 22. 19. 23. 21.]
 [23. 24. 18. 19. 22. 23. 21. 18. 23. 15. 25. 21.]
 [18. 21. 21. 18. 26. 24. 18. 17. 18. 20. 21. 20.]
 [19. 12. 21. 19. 23. 20. 22. 19. 17. 19. 20. 24.]
 [18. 22. 24. 17. 21. 28. 18. 16. 22. 24. 25. 24.]]


In [6]:
M, N= data_txtfile.shape
print(M) #no of rows(factories)
print(N) #no of columns(location)

12
12


In [7]:
# create a model
model_lab6_ex2 = ConcreteModel()

In [8]:
row_indices = range(M)
col_indices = range(N)

print(row_indices)
print(col_indices)

range(0, 12)
range(0, 12)


In [33]:
#to get the contents of costs alone from the file we use the following
for row in row_indices:
  for col in col_indices:
    print('cost[',row,'][',col,']:',data_txtfile[row, col])

cost[ 0 ][ 0 ]: 21.0
cost[ 0 ][ 1 ]: 25.0
cost[ 0 ][ 2 ]: 18.0
cost[ 0 ][ 3 ]: 19.0
cost[ 0 ][ 4 ]: 23.0
cost[ 0 ][ 5 ]: 24.0
cost[ 0 ][ 6 ]: 16.0
cost[ 0 ][ 7 ]: 17.0
cost[ 0 ][ 8 ]: 21.0
cost[ 0 ][ 9 ]: 23.0
cost[ 0 ][ 10 ]: 19.0
cost[ 0 ][ 11 ]: 21.0
cost[ 1 ][ 0 ]: 19.0
cost[ 1 ][ 1 ]: 22.0
cost[ 1 ][ 2 ]: 19.0
cost[ 1 ][ 3 ]: 19.0
cost[ 1 ][ 4 ]: 23.0
cost[ 1 ][ 5 ]: 21.0
cost[ 1 ][ 6 ]: 15.0
cost[ 1 ][ 7 ]: 20.0
cost[ 1 ][ 8 ]: 16.0
cost[ 1 ][ 9 ]: 17.0
cost[ 1 ][ 10 ]: 17.0
cost[ 1 ][ 11 ]: 24.0
cost[ 2 ][ 0 ]: 21.0
cost[ 2 ][ 1 ]: 19.0
cost[ 2 ][ 2 ]: 20.0
cost[ 2 ][ 3 ]: 10000.0
cost[ 2 ][ 4 ]: 22.0
cost[ 2 ][ 5 ]: 24.0
cost[ 2 ][ 6 ]: 14.0
cost[ 2 ][ 7 ]: 16.0
cost[ 2 ][ 8 ]: 18.0
cost[ 2 ][ 9 ]: 16.0
cost[ 2 ][ 10 ]: 20.0
cost[ 2 ][ 11 ]: 24.0
cost[ 3 ][ 0 ]: 19.0
cost[ 3 ][ 1 ]: 22.0
cost[ 3 ][ 2 ]: 17.0
cost[ 3 ][ 3 ]: 17.0
cost[ 3 ][ 4 ]: 22.0
cost[ 3 ][ 5 ]: 22.0
cost[ 3 ][ 6 ]: 18.0
cost[ 3 ][ 7 ]: 18.0
cost[ 3 ][ 8 ]: 20.0
cost[ 3 ][ 9 ]: 10000.0
cost[ 3 ][ 10 ]: 23.0


In [10]:
#declare the decision variables in the model
model_lab6_ex2.x = Var(row_indices, col_indices, domain= NonNegativeIntegers)

In [11]:
model_lab6_ex2.objective = Objective(expr=sum(data_txtfile[i,j]*model_lab6_ex2.x[i,j] for i in row_indices for j in col_indices),sense=minimize)

In [12]:
model_lab6_ex2.constraints = ConstraintList()

In [13]:
#adding constraints of locating one factory at each location and assigning one location for each factory
for i in range(M):
  model_lab6_ex2.constraints.add(expr=sum(model_lab6_ex2.x[i,j] for j in col_indices) == 1)

for j in range(N):
  model_lab6_ex2.constraints.add(expr=sum(model_lab6_ex2.x[i,j] for i in row_indices) == 1)

In [14]:
model_lab6_ex2.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   24 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24}
    x_index : Size=1, Index=None, Ordered=True
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :  144 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10), (0, 11), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10), (5, 11

In [15]:
!apt-get install -y -qq coinor-cbc

In [16]:
opt_cbc = SolverFactory('cbc')

In [17]:
result = opt_cbc.solve(model_lab6_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [18]:
# display solution
print('\nObjective = ', model_lab6_ex2.objective())

for i in row_indices:
  for j in col_indices:
    if model_lab6_ex2.x[i,j].value==1:
      print('Factory', i+1,'is to be located at Location', j+1)

print('\nConstraints')
model_lab6_ex2.constraints.display()


Objective =  204.0
Factory 1 is to be located at Location 11
Factory 2 is to be located at Location 9
Factory 3 is to be located at Location 8
Factory 4 is to be located at Location 6
Factory 5 is to be located at Location 3
Factory 6 is to be located at Location 12
Factory 7 is to be located at Location 7
Factory 8 is to be located at Location 4
Factory 9 is to be located at Location 10
Factory 10 is to be located at Location 1
Factory 11 is to be located at Location 2
Factory 12 is to be located at Location 5

Constraints
constraints : Size=24
    Key : Lower : Body : Upper
      1 :   1.0 :  1.0 :   1.0
      2 :   1.0 :  1.0 :   1.0
      3 :   1.0 :  1.0 :   1.0
      4 :   1.0 :  1.0 :   1.0
      5 :   1.0 :  1.0 :   1.0
      6 :   1.0 :  1.0 :   1.0
      7 :   1.0 :  1.0 :   1.0
      8 :   1.0 :  1.0 :   1.0
      9 :   1.0 :  1.0 :   1.0
     10 :   1.0 :  1.0 :   1.0
     11 :   1.0 :  1.0 :   1.0
     12 :   1.0 :  1.0 :   1.0
     13 :   1.0 :  1.0 :   1.0
     14 :   1

In [19]:
#changing domain of variables to continuous values
for i in row_indices:
  for j in col_indices:
    model_lab6_ex2.x[i,j].domain=NonNegativeReals

In [20]:
result = opt_cbc.solve(model_lab6_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [21]:
# display solution
print('\nObjective = ', model_lab6_ex2.objective())

for i in row_indices:
  for j in col_indices:
    if model_lab6_ex2.x[i,j].value==1:
      print('Factory', i+1,'is to be located at Location', j+1)

print('\nConstraints')
model_lab6_ex2.constraints.display()


Objective =  204.0
Factory 1 is to be located at Location 11
Factory 2 is to be located at Location 9
Factory 3 is to be located at Location 8
Factory 4 is to be located at Location 6
Factory 5 is to be located at Location 3
Factory 6 is to be located at Location 12
Factory 7 is to be located at Location 7
Factory 8 is to be located at Location 4
Factory 9 is to be located at Location 10
Factory 10 is to be located at Location 1
Factory 11 is to be located at Location 2
Factory 12 is to be located at Location 5

Constraints
constraints : Size=24
    Key : Lower : Body : Upper
      1 :   1.0 :  1.0 :   1.0
      2 :   1.0 :  1.0 :   1.0
      3 :   1.0 :  1.0 :   1.0
      4 :   1.0 :  1.0 :   1.0
      5 :   1.0 :  1.0 :   1.0
      6 :   1.0 :  1.0 :   1.0
      7 :   1.0 :  1.0 :   1.0
      8 :   1.0 :  1.0 :   1.0
      9 :   1.0 :  1.0 :   1.0
     10 :   1.0 :  1.0 :   1.0
     11 :   1.0 :  1.0 :   1.0
     12 :   1.0 :  1.0 :   1.0
     13 :   1.0 :  1.0 :   1.0
     14 :   1

In [22]:
#changing the cost to non-integer values
for i in row_indices:
  for j in col_indices:
    data_txtfile[i,j]=data_txtfile[i,j]/10
print(data_txtfile)

[[2.1 2.5 1.8 1.9 2.3 2.4 1.6 1.7 2.1 2.3 1.9 2.1]
 [1.9 2.2 1.9 1.9 2.3 2.1 1.5 2.  1.6 1.7 1.7 2.4]
 [2.1 1.9 2.  1.8 2.2 2.4 1.4 1.6 1.8 1.6 2.  2.4]
 [1.9 2.2 1.7 1.7 2.2 2.2 1.8 1.8 2.  2.1 2.3 2.3]
 [1.8 2.2 1.5 1.5 2.3 2.3 2.1 2.5 2.4 1.9 2.1 2.3]
 [2.3 1.9 1.9 1.4 2.4 2.1 2.3 2.1 2.  2.  2.2 1.9]
 [2.1 1.6 2.7 1.7 2.6 2.7 1.6 1.9 2.4 2.  2.  2.3]
 [2.2 1.7 2.1 1.5 2.3 2.4 2.4 2.  2.2 1.9 2.3 2.1]
 [2.3 2.4 1.8 1.9 2.2 2.3 2.1 1.8 2.3 1.5 2.5 2.1]
 [1.8 2.1 2.1 1.8 2.6 2.4 1.8 1.7 1.8 2.  2.1 2. ]
 [1.9 1.2 2.1 1.9 2.3 2.  2.2 1.9 1.7 1.9 2.  2.4]
 [1.8 2.2 2.4 1.7 2.1 2.8 1.8 1.6 2.2 2.4 2.5 2.4]]


In [23]:
model_lab6_ex2.objective = Objective(expr=sum(data_txtfile[i,j]*model_lab6_ex2.x[i,j] for i in row_indices for j in col_indices),sense=minimize)

This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


In [24]:
model_lab6_ex2.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   24 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24}
    x_index : Size=1, Index=None, Ordered=True
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :  144 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10), (0, 11), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10), (5, 11

In [25]:
result = opt_cbc.solve(model_lab6_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [37]:
#display solution
#we observe that values of x_ij still remain integral valued(0 or 1 here because of constraints)

print('\nObjective = ', model_lab6_ex2.objective())

for i in row_indices:
  for j in col_indices:
    print("Value of x[",i,"][",j,"] is", model_lab6_ex2.x[i,j].value)

for i in row_indices:
  for j in col_indices:
    if model_lab6_ex2.x[i,j].value==1:
      print('Factory', i+1,'is to be located at Location', j+1)
      
print('\nConstraints')
model_lab6_ex2.constraints.display()


Objective =  204.0
Value of x[ 0 ][ 0 ] is 0.0
Value of x[ 0 ][ 1 ] is 0.0
Value of x[ 0 ][ 2 ] is 0.0
Value of x[ 0 ][ 3 ] is 0.0
Value of x[ 0 ][ 4 ] is 0.0
Value of x[ 0 ][ 5 ] is 0.0
Value of x[ 0 ][ 6 ] is 0.0
Value of x[ 0 ][ 7 ] is 0.0
Value of x[ 0 ][ 8 ] is 0.0
Value of x[ 0 ][ 9 ] is 0.0
Value of x[ 0 ][ 10 ] is 1.0
Value of x[ 0 ][ 11 ] is 0.0
Value of x[ 1 ][ 0 ] is 0.0
Value of x[ 1 ][ 1 ] is 0.0
Value of x[ 1 ][ 2 ] is 0.0
Value of x[ 1 ][ 3 ] is 0.0
Value of x[ 1 ][ 4 ] is 0.0
Value of x[ 1 ][ 5 ] is 0.0
Value of x[ 1 ][ 6 ] is 0.0
Value of x[ 1 ][ 7 ] is 0.0
Value of x[ 1 ][ 8 ] is 1.0
Value of x[ 1 ][ 9 ] is 0.0
Value of x[ 1 ][ 10 ] is 0.0
Value of x[ 1 ][ 11 ] is 0.0
Value of x[ 2 ][ 0 ] is 0.0
Value of x[ 2 ][ 1 ] is 0.0
Value of x[ 2 ][ 2 ] is 0.0
Value of x[ 2 ][ 3 ] is 0.0
Value of x[ 2 ][ 4 ] is 0.0
Value of x[ 2 ][ 5 ] is 0.0
Value of x[ 2 ][ 6 ] is 0.0
Value of x[ 2 ][ 7 ] is 1.0
Value of x[ 2 ][ 8 ] is 0.0
Value of x[ 2 ][ 9 ] is 0.0
Value of x[ 2 ][ 10 ] is

In [27]:
#reseting the cost to previous integer values
for i in row_indices:
  for j in col_indices:
    data_txtfile[i,j]=data_txtfile[i,j]*10

#restrictions on factory allocation at specific locations
data_txtfile[2,3]=10000
data_txtfile[8,6]=10000
data_txtfile[3,9]=10000

print(data_txtfile)

[[   21.    25.    18.    19.    23.    24.    16.    17.    21.    23.
     19.    21.]
 [   19.    22.    19.    19.    23.    21.    15.    20.    16.    17.
     17.    24.]
 [   21.    19.    20. 10000.    22.    24.    14.    16.    18.    16.
     20.    24.]
 [   19.    22.    17.    17.    22.    22.    18.    18.    20. 10000.
     23.    23.]
 [   18.    22.    15.    15.    23.    23.    21.    25.    24.    19.
     21.    23.]
 [   23.    19.    19.    14.    24.    21.    23.    21.    20.    20.
     22.    19.]
 [   21.    16.    27.    17.    26.    27.    16.    19.    24.    20.
     20.    23.]
 [   22.    17.    21.    15.    23.    24.    24.    20.    22.    19.
     23.    21.]
 [   23.    24.    18.    19.    22.    23. 10000.    18.    23.    15.
     25.    21.]
 [   18.    21.    21.    18.    26.    24.    18.    17.    18.    20.
     21.    20.]
 [   19.    12.    21.    19.    23.    20.    22.    19.    17.    19.
     20.    24.]
 [   18.    22.    24

In [28]:
#reseting domain back to non negative integers
for i in row_indices:
  for j in col_indices:
    model_lab6_ex2.x[i,j].domain=NonNegativeIntegers

In [29]:
model_lab6_ex2.objective = Objective(expr=sum(data_txtfile[i,j]*model_lab6_ex2.x[i,j] for i in row_indices for j in col_indices),sense=minimize)

This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


In [30]:
model_lab6_ex2.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   24 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24}
    x_index : Size=1, Index=None, Ordered=True
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :  144 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10), (0, 11), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10), (5, 11

In [31]:
result = opt_cbc.solve(model_lab6_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [32]:
# display solution
print('\nObjective = ', model_lab6_ex2.objective())

for i in row_indices:
  for j in col_indices:
    if model_lab6_ex2.x[i,j].value==1:
      print('Factory', i+1,'is to be located at Location', j+1)

print('\nConstraints')
model_lab6_ex2.constraints.display()


Objective =  204.0
Factory 1 is to be located at Location 11
Factory 2 is to be located at Location 9
Factory 3 is to be located at Location 8
Factory 4 is to be located at Location 6
Factory 5 is to be located at Location 3
Factory 6 is to be located at Location 12
Factory 7 is to be located at Location 7
Factory 8 is to be located at Location 4
Factory 9 is to be located at Location 10
Factory 10 is to be located at Location 1
Factory 11 is to be located at Location 2
Factory 12 is to be located at Location 5

Constraints
constraints : Size=24
    Key : Lower : Body : Upper
      1 :   1.0 :  1.0 :   1.0
      2 :   1.0 :  1.0 :   1.0
      3 :   1.0 :  1.0 :   1.0
      4 :   1.0 :  1.0 :   1.0
      5 :   1.0 :  1.0 :   1.0
      6 :   1.0 :  1.0 :   1.0
      7 :   1.0 :  1.0 :   1.0
      8 :   1.0 :  1.0 :   1.0
      9 :   1.0 :  1.0 :   1.0
     10 :   1.0 :  1.0 :   1.0
     11 :   1.0 :  1.0 :   1.0
     12 :   1.0 :  1.0 :   1.0
     13 :   1.0 :  1.0 :   1.0
     14 :   1

**Answer 1:**

$minimize \sum_{i=1}^n\sum_{j=1}^n C(i,j)*x_{ij} $

subject to constraints

$\sum_{i=1}^nx_{ij}=1\ \forall i \in \{1,2,....n\}$

$\sum_{j=1}^nx_{ij}=1\ \forall j \in \{1,2,....n\}$

$ x_{ij} \geq 0\ \forall\ i,j \in \{1,2,....,n\}$

where C(i,j) is the cost of locating factory i at location j. $x_{ij}$ indicates whether factory i is located at location j or not, for which the values are 1 and 0 respectively (since they are taken to be non negative integers and the first two constraints restricts them from taking values more than 1).

**Answer 8:**

Optimal cost =  204.0

Factory 1 is to be located at Location 11

Factory 2 is to be located at Location 9

Factory 3 is to be located at Location 8

Factory 4 is to be located at Location 6

Factory 5 is to be located at Location 3

Factory 6 is to be located at Location 12

Factory 7 is to be located at Location 7

Factory 8 is to be located at Location 4

Factory 9 is to be located at Location 10

Factory 10 is to be located at Location 1

Factory 11 is to be located at Location 2

Factory 12 is to be located at Location 5

**Answer 9:**

After changing the variables to be non negative reals from n negative integers

Optimal cost =  204.0

Factory 1 is to be located at Location 11

Factory 2 is to be located at Location 9

Factory 3 is to be located at Location 8

Factory 4 is to be located at Location 6

Factory 5 is to be located at Location 3

Factory 6 is to be located at Location 12

Factory 7 is to be located at Location 7

Factory 8 is to be located at Location 4

Factory 9 is to be located at Location 10

Factory 10 is to be located at Location 1

Factory 11 is to be located at Location 2

Factory 12 is to be located at Location 5

**Answer 10:**

Yes, the optimal costs for both the problems are same and they are still integer valued. If $x_{ij}$'s took fractional values (which add upto 1 in each row and column of ($x_{ij}$) matrix) then the solution would not have been optimal as the total cost (sum of C(i,j)*$x_{ij}$) would have been greater than when compared to taking only one $x_{ij}$ value as 1 in each row and column.

**Answer 12:**

The changes I have made are increasing the cost of locating facilities 3, 9 and 4 at location 4, 6 and 10 respectively to 10000. Since the objective is to minimize total cost, $x_{ij}$ for them will be minimum (which is 0). So, by making these changes we can solve the assignment problem keeping in mind the three new restrictions.

Optimal cost =  204.0

Factory 1 is to be located at Location 11

Factory 2 is to be located at Location 9

Factory 3 is to be located at Location 8

Factory 4 is to be located at Location 6

Factory 5 is to be located at Location 3

Factory 6 is to be located at Location 12

Factory 7 is to be located at Location 7

Factory 8 is to be located at Location 4

Factory 9 is to be located at Location 10

Factory 10 is to be located at Location 1

Factory 11 is to be located at Location 2

Factory 12 is to be located at Location 5

Previously also when we got the optimal solution, facility 3 was not assigned to location 4, facility 9 was not assigned to location 6 and facility 4 was not assigned to location 10. So, there are no new changes and the optimal cost remains to be same (204).